bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryFunctor.scala

```diff
@@ -0,0 +1,230 @@
+package edu.gemini.lchquery.servlet
+
+import java.io.Serializable
+import java.security.Principal
+import java.text.SimpleDateFormat
+import java.util
+import java.util.{Date, TimeZone}
+import java.util.logging.{Level, Logger}
+
+import edu.gemini.odb.browser._
+import edu.gemini.pot.sp.{ISPNode, ISPObservation, ISPProgram}
+import edu.gemini.pot.spdb.{DBAbstractQueryFunctor, IDBDatabaseService, IDBFunctor, IDBParallelFunctor}
+import edu.gemini.skycalc.{DDMMSS, HHMMSS}
+import edu.gemini.spModel.gemini.obscomp.SPSiteQuality
+import edu.gemini.spModel.obs.{ObsClassService, ObservationStatus}
+import edu.gemini.spModel.rich.shared.immutable._
+import edu.gemini.spModel.target.SPTarget
+import edu.gemini.spModel.target.env.TargetEnvironment
+import edu.gemini.spModel.too.Too
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+
+class LchQueryFunctor(queryType: LchQueryFunctor.QueryType,
+                      programParams: List[(LchQueryParam[ISPProgram], String)],
+                      observationParams: List[(LchQueryParam[ISPObservation], String)])
+  extends DBAbstractQueryFunctor with IDBParallelFunctor {
+
+  // The current QueryResult. Must be mutable.
+  var queryResult: QueryResult = new QueryResult() {
+    setProgramsNode(new ProgramsNode())
+  }
+
+  // Called once per program by IDBQueryRunner implementation.
+  override def execute(db: IDBDatabaseService, node: ISPNode, principals: util.Set[Principal]): Unit = {
+    val prog = node.asInstanceOf[ISPProgram]
+
+    def programMatches: Boolean = programParams.forall {
+      case (LchQueryParam(_, valueMatcher), paramValue) => valueMatcher.matches(paramValue, prog)
+    }
+    def observationMatches(obs: ISPObservation): Boolean = observationParams.forall {
+      case (LchQueryParam(_, valueMatcher), paramValue) => valueMatcher.matches(paramValue, obs)
+    }
+
+    Try {
+      // See if an ISPProgram matches the query specifications.
+      if (programMatches) {
+        val matchingObs = prog.getAllObservations.asScala.toList.filter(observationMatches)
+        if (matchingObs.nonEmpty) {
+          addProgram(prog, if (queryType == LchQueryFunctor.QueryType.ProgramQuery) Nil else matchingObs)
+        }
+      }
+    } match {
+      case Success(_) =>
+      case Failure(t) =>
+        LchQueryFunctor.Log.log(Level.SEVERE, "problem running LchQueryFunctor", t)
+        throw new RuntimeException(t)
+    }
```

Comment:
We do the exception wrapping in a number of different ways in the codebase, and I've personally contributed to doing it in every different way possible.  These days though I've settled on using `\/.fromTryCatchNonFatal` since we make extensive use of `\/` anyway.  Does everyone agree it would be good to handle this in the same way everywhere and that `\/` should be that way?  An argument against is that eventually presumably we'll replace scalaz `\/` with cats `Xor` or something.  Still that will be a more or less mechanical switch that we have to do anyway.